Add accessibility documentation for heading components (#1294)
sumup-oss/circuit-ui: packages/circuit-ui/components/Headline/Headline.docs.mdx

<Status.Stable />

Headlines are used for titling sections of the interface. They're an essential
part of a page's structure, helping organize information for our users.

<Story id="typography-headline--base" />
<Props />

## Component variations

### Sizes

The Headline component comes in four sizes.

Use the `four` size for card headers and `three` for page titles in web
applications. For specific use cases such as landing pages, consider using the
[Title](Typography/Title) component.

<Story id="typography-headline--sizes" />

## Accessibility

Headings are a fundamental aspect of creating accessible interfaces. Sighted
users rely on them to easily find what they want on a page, while assistive
technology allows users to jump between them for easier access to the page's
different sections.

### Best practices

#### Each page should have a single `<h1>`

> The `<h1>` represents the page title. Our pages should all have one (and only

Reviewer: I would avoid using "our", "we", or "us" when writing docs. At best, it's ambiguous (who's "we"?), at worst, it's exclusionary (I don't belong to "us"). 

```suggestion
The `<h1>` represents the page title. Pages should all have one (and only
```

one), and it should be similar to the document's `<title>`.

#### Headings should be descriptive

A heading's text should describe its section's content accurately and concisely.
Because screen reader users will use them to navigate the page's contents,
headings must also make sense out of the content's context, as if rendered in a
table of contents.

> #### Headings should be nested incrementally

Reviewer: ```suggestion
#### Headings should be nested hierarchically
```

Beyond the `<h1>`, heading levels should be nested incrementally to make them
easily accessible to users. Think of your heading structure like a nested list:

- `<h1>` Page title
  - `<h2>` Section title
    - `<h3>` Section title
    - `<h3>` Section title
      - `<h4>` Section title
  - `<h2>` Section title

Don't skip heading levels: an `<h4>` should not directly follow an `<h2>`.

In contrast, skipping headline _sizes_ and jumping from e.g.

> `<Headline size="two" as="h2" />` to `<Headline size="four" as="h4" />` does not

Reviewer: Sizes can be skipped, not levels 😉

```suggestion
`<Headline size="two" as="h2" />` to `<Headline size="four" as="h3" />` does not
```

Author: Oopsies

Author: 😂

constitute an accessibility issue, although we recommend keeping sizes visually
consistent across pages.

#### Avoid sections without headings

If a section doesn't have a heading (for example, if it uses a different
background color to distinguish it from another section), it will not be easily
accessible to blind users.

Consider adding a heading to make the section's purpose clearer.

You can also add visually hidden headings using the `hideVisually()` style
mixin:

```tsx
import { Headline, hideVisually } from '@sumup/circuit-ui';

function HiddenHeadline() {
  return (
    <Headline as="h2" css={hideVisually}>
      My section
    </Headline>
  );
}
```

#### Avoid headings without sections

> Headings typically use large typography, but all large typography shoulnd't be a

Reviewer: ```suggestion
Headings typically use large typography, but not all large typography should be a
```

heading.

If you need to render large text that doesn't describe a section, don't use a
semantic heading element, and instead use CSS to enlarge the text.

```tsx
<section>
  <Headline as="h2" size="four">
    My product
  </Headline>

>   <Headline as="p" size="one">

Author: We don't support this right now. Should we? Is there a reason not to? If we won't support this, how do we cater to the use case of a developer wanting to use large typography without a semantic heading element?

Author: cc @connor-baer—thoughts?

Reviewer: While there is no strict coupling between size and heading level, they should be roughly aligned. Larger text communicates importance to sighted users and thus should also be reflected semantically for visually-impaired users.

I can only think of a single use case for large text that's not a headline: the price tags on the marketing website. IMO, a custom styled component makes more sense there. **Edit**: I just realized that you've used a price as the example here 😃

Author: Sounds good, I'll document this. Maybe we can recommend using the typography values from the theme

    9,99 €
  </Headline>
  <Body>A description of my product.</Body>
</section>
```

### Resources

#### Testing heading acessibility

Use a tool like [Wave](https://wave.webaim.org/) to extract structure from a
page (in Wave, you'll find the page structure under the _Structure_ tab). Verify
that sections are appropriately titled and are nested incrementally.

#### Related WCAG success criteria

- 1.3.1:
  [Info and Relationships](https://www.w3.org/WAI/WCAG21/Understanding/info-and-relationships.html)
- 2.4.6:

>   [Headings and Labels](https://www.w3.org/WAI/WCAG21/Understanding/headings-and-labels.html)

Reviewer: I love this!
